src/api/spec/controllers/webui/obs_factory/distributions_controller_spec.rb

```diff
@@ -0,0 +1,77 @@
+require 'rails_helper'
+require 'webmock/rspec'
+
+RSpec.describe Webui::ObsFactory::DistributionsController, type: :controller, vcr: true do
+  let(:factory) { create(:project, name: 'openSUSE:Factory') }
+  let!(:factory_staging_a) { create(:project, name: 'openSUSE:Factory:Staging:A', description: 'Factory staging project A') }
+  let!(:factory_ring_bootstrap) { create(:project, name: 'openSUSE:Factory:Rings:0-Bootstrap', description: 'Factory ring project') }
+  let(:target_package) { create(:package, name: 'target_package', project: factory) }
+  let(:source_project) { create(:project, name: 'source_project') }
+  let(:source_package) { create(:package, name: 'source_package', project: source_project) }
+
+  describe 'GET #show' do
+    before do
+      allow(::ObsFactory::OpenqaJob).to receive(:find_all_by).and_return([:fake, :content])
+
+      get :show, params: { project: factory.name }
+    end
+
+    it { expect(response).to have_http_status(:success) }
+    it { expect(response).to render_template(:show) }
+
+    it { expect(assigns(:staging_projects).first.project).to eq(factory_staging_a) }
```

Comment:
I would like to ask a newbie question: why make this so many it{}s? This takes like forever and I see no additional value over one it with multiple expects.

Reply:
We try to follow the "one expectation per it"  rule, the point is to have a failing expect in each example and not having one failure covering the subsequent expects failures.

If you just put all the expects in one it and the first one fails you won't see the other failing. You would have to fix the first and after rerun the test and see the next one failing... having one expect per it you will see all the failures at the same time and maybe you realize of the problem earlier.

But, in opposite to that, you can have multiple expect in a it if there is a big performance pain in the setup of the test...

See http://www.betterspecs.org/#single, we started to follow betterspecs.org "rules" time ago and seems that is a good help.

Reply:
I guess you didn't scroll down on that page, because:

"Deeply test your models and your application behaviour (integration tests). Do not add useless complexity testing controllers."

:laughing: :laughing: :laughing:

Reply:
@coolo if you continue scrolling down:

> This is an open debate in the Ruby community and both sides have good arguments supporting their idea. People supporting the need of testing controllers will tell you that your integration tests don't cover all use cases and that they are slow.

We are testing controllers... I think it is not something we want to discuss here... It is true that this is maybe testing too much... but now that it is done... I only see advantage in keeping it. :smile: